presto-main/src/main/java/io/prestosql/sql/planner/iterative/rule/DetermineJoinDistributionType.java

```diff
@@ -91,22 +98,87 @@ public static boolean canReplicate(JoinNode joinNode, Context context)
         return buildSideSizeInBytes <= joinMaxBroadcastTableSize.toBytes();
     }
 
+    public static double getSourceTablesSizeInBytes(PlanNode node, Context context)
+    {
+        return getSourceTablesSizeInBytes(node, context.getLookup(), context.getStatsProvider(), context.getSymbolAllocator().getTypes());
+    }
+
+    @VisibleForTesting
+    static double getSourceTablesSizeInBytes(PlanNode node, Lookup lookup, StatsProvider statsProvider, TypeProvider typeProvider)
+    {
+        boolean hasExpandingNodes = PlanNodeSearcher.searchFrom(node, lookup)
+                .where(isInstanceOfAny(JoinNode.class))
+                .matches();
+        if (hasExpandingNodes) {
+            return Double.NaN;
+        }
+
+        List<PlanNode> sourceNodes = PlanNodeSearcher.searchFrom(node, lookup)
+                .where(isInstanceOfAny(TableScanNode.class, ValuesNode.class))
+                .findAll();
+
+        return sourceNodes.stream()
+                .mapToDouble(sourceNode -> statsProvider.getStats(sourceNode).getOutputSizeInBytes(sourceNode.getOutputSymbols(), typeProvider))
+                .sum();
+    }
+
     private PlanNode getCostBasedJoin(JoinNode joinNode, Context context)
     {
         List<PlanNodeWithCost> possibleJoinNodes = new ArrayList<>();
 
         addJoinsWithDifferentDistributions(joinNode, possibleJoinNodes, context);
         addJoinsWithDifferentDistributions(joinNode.flipChildren(), possibleJoinNodes, context);
+        boolean containsReplicatedJoin = possibleJoinNodes.stream()
+                .map(PlanNodeWithCost::getPlanNode)
+                .map(JoinNode.class::cast)
+                .anyMatch(node -> node.getDistributionType().equals(Optional.of(REPLICATED)));
+
+        JoinNode sizedBasedJoin = getSizeBasedJoin(joinNode, context);
+        if (!containsReplicatedJoin && sizedBasedJoin.getDistributionType().equals(Optional.of(REPLICATED))) {
+            // choose replicated size based join if stats calculator would skip replicated join altogether
+            // (because of inaccurate or missing intermediate plan node stats)
+            return sizedBasedJoin;
```

Comment:
`containsReplicatedJoin` seems straighforward, based on `addJoinsWithDifferentDistributions` -- there is no cost decision yet.

i am concerned we exit too early here. 
Would it be sufficient to add this to `if (possibleJoinNodes.stream().anyMatch(result -> result.getCost().hasUnknownComponents()) || possibleJoinNodes.isEmpty())` branch below?